pkg/ring/model.go

```diff
@@ -442,3 +642,264 @@ func (d *Desc) RemoveTombstones(limit time.Time) {
+		if _, ok := distinct[predecessor.Ingester]; !ok && len(distinct) == opts.Offset+1 {
+			break
+		} else if !IsHealthyState(&ing, predecessor.State, opts.Op, opts.MaxHeartbeat) {
+			continue
+		}
+
+		// Collect the token if its successor is our starting token.
+		succ, err := d.successor(NeighborOptions{
+			Start:        predecessor.StatefulToken(),
+			Offset:       opts.Offset,
+			Op:           opts.Op,
+			MaxHeartbeat: opts.MaxHeartbeat,
+			IncludeStart: true,
+		}, healthyOrStart)
+		if err != nil {
+			return predecessors, err
+		} else if succ.Token == opts.Start.Token {
+			predecessors = append([]TokenDesc{predecessor}, predecessors...)
+		}
+
+		distinct[predecessor.Ingester] = struct{}{}
+	}
+
+	return predecessors, nil
+}
+
+// Successor moves around the ring to find a neighbor of opts.Start.
+// The healthiness of the ingester and the state of the token defines
+// which tokens are considered as successors.
+//
+// Tokens are only considered once based on their ingester; if
+// opts.IncludeStart is true, then the ingster for the opts.Start
+// argument will not be considered as a neighbor (i.e., it will act
+// as if we've already seen it).
+//
+// If opts.Offset has a positive value, Successor searches the ring
+// clockwise. Otherwise, if opts.Offset has a negative value, Successor
+// will search the ring counter-clockwise. An offset of 0 will return
+// the starting token.
+func (d *Desc) Successor(opts NeighborOptions) (TokenDesc, error) {
+	return d.successor(opts, func(t TokenDesc, s State, o Operation, hb time.Duration) bool {
+		ing := d.Ingesters[t.Ingester]
+		return IsHealthyState(&ing, s, o, hb)
+	})
+}
+
+type healthCheckFunc func(TokenDesc, State, Operation, time.Duration) bool
+
+func (d *Desc) successor(opts NeighborOptions, healthy healthCheckFunc) (TokenDesc, error) {
+	idx := d.search(opts.Start.Token)
+	if d.Tokens[idx].Token != opts.Start.Token {
+		return TokenDesc{}, fmt.Errorf("could not find token %d in ring", opts.Start.Token)
+	}
+	if opts.Offset == 0 {
+		return d.Tokens[idx], nil
+	}
+
+	numSuccessors := opts.Offset
+	if numSuccessors < 0 {
+		numSuccessors = -numSuccessors
+	}
+
+	successors := make([]TokenDesc, 0, numSuccessors)
+	distinct := map[string]struct{}{}
+	if opts.IncludeStart {
+		distinct[d.Tokens[idx].Ingester] = struct{}{}
+	}
+
+	startIdx := idx
+
+	direction := 1
+	if opts.Offset < 0 {
+		direction = -1
+	}
+
+	for {
+		idx += direction
+		if idx < 0 {
+			idx = len(d.Tokens) - 1
+		} else {
+			idx %= len(d.Tokens)
+		}
+
+		// Stop if we've completely circled the ring
+		if idx == startIdx {
+			break
+		}
+
+		successor := d.Tokens[idx]
+		if _, ok := distinct[successor.Ingester]; ok {
+			continue
+		}
+
+		if healthy(successor, successor.State, opts.Op, opts.MaxHeartbeat) {
+			successors = append(successors, successor)
+			if len(successors) == numSuccessors {
+				return successor, nil
+			}
+
+			distinct[successor.Ingester] = struct{}{}
+		}
+	}
```

Comment:
I suspect both `Predecessors` and `Successors` would be slightly easier to read if they called upon a ring iterator that abstracted out handling token indices, health checks, and wrapping around the ring.